### PR TITLE
RSDK-9173: Data API Go SDK

### PR DESCRIPTION
**Pull request by jckras**
This PR adds the Data API wrappers to the Go SDK

### CODE REVIEW

File: app/data_test.go
Comment (purplenicole730) [inline app/data_test.go:339]: Why not just test if it's equal to `organizationID` directly? This is a question that applies to the other expected variables as well.

Comment (jckras) [reply app/data_test.go:339]: I was doing it this way because in this test we are expecting to get back the protobuf version of that variable and sometimes the protobuf version is the same thing (like when it is a string or int) but other times it is different and I would pass the original variable through a "ToProto" function. So, because I was sometimes passing it thru a function and othertimes not, I was just trying to stay consistent. I also did this to help me understand exactly what I was testing but I can def change it if its repetitive or makes it more unclear.

Comment (purplenicole730) [reply app/data_test.go:339]: It is usually bad practice to create a string that's just a repeat and is already pretty well defined. It adds repetition and unnecessary confusion. I do understand what you mean though about consistency and such, but strings and ints should already be clear.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:268]: I noticed that this function is only called in the tests. I also did something similar originally, but I realized that `go-staticcheck` did not like that (see below), and I agree.
image
The purpose of the main code is to provide code functionality, while testing is supposed to test said functionality. Testing should be isolated, and writing functions that will only be used in tests is not really adding to code functionality. I'd suggest putting this function in `data_test.go`.

File: app/data_test.go
Comment (purplenicole730) [inline app/data_test.go:1]: Just for consistency, I would name this `data_client_test.go`

File: app/data_client.go
Comment (purplenicole730) [reply app/data_client.go:268]: This also applies to the other private functions that are only used in the test

**Review by purplenicole730 — CHANGES_REQUESTED** [review]
It's looking good! Some suggestions on how to approach testing

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:503]: here (and below), can we update the comment to give clarity on what the return value means? I don't think it's intuitive/obvious that we're returning a count of how much data was deleted.

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:547]:
```suggestion
	return err
```

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:559]:
```suggestion
	return err
```

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:602]: I think we need a bit more clarification in the docstring on how this works. the float values, e.g., _must_ be somewhere between 0 and 1, but that's not obvious based on what we have here. I don't think we need as much detail as exists in the python SDK, but taking a peek there might be useful for more context to provide.

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:701]:
```suggestion
	return err
```

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:717]:
```suggestion
	return err
```

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:733]:
```suggestion
	return err
```

**Review by stuqdog — CHANGES_REQUESTED** [review]
Nice, got some suggestions (most of them minor) but generally this is looking quite good!

File: app/data_client.go
Comment (jckras) [reply app/data_client.go:412]: Oh good catch thank you!

**Review by purplenicole730 — COMMENTED** [review]
Holding off on giving the final approval until viam_client changes are also in!

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:266]: We should delete this

File: app/data_client.go
Comment (stuqdog) [inline app/data_client.go:264]: Can we put this in `protoutils`? It's a pretty generic function, and while `data_client` is the only current user, I suspect there may be broader use in the future (much like `ConvertStringMapToAnyPBMap`).

**Review by stuqdog — COMMENTED** [review]
A couple small changes, and also we should update the `UpdateBoundingBox` comment to clarify the acceptable float ranges. Other than that, lgtm!

**Review by stuqdog — APPROVED** [review]
lgtm modulo change to `NewDataClient` that we discussed offline :)

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:180]: Just wanted to note that it says that there's a logger, but there is no logger mentioned in the function.

Comment (purplenicole730) [inline app/data_client.go:180]: Just wanted to note that it says that there's a logger, but there is no logger mentioned in the function.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:187]: `err` seems to always be `nil`, so there should only be one return type here: `*DataClient`.

**Review by purplenicole730 — APPROVED** [review]
Some comments about adding to ViamClient, but everything else LGTM!

File: app/data_client.go
Comment (purplenicole730) [reply app/data_client.go:187]: And then when this change is made, make sure to remove the error return in `ViamClient.DataClient()` as well.

**Pull request merged**
